docker_* modules: simplify idempotency comparisons

felixfontein opened the pull request:
##### SUMMARY
docker_container used to contain a framework for doing comparisons; we refactored that to fix some idempotency bugs (#45905, #44789). I propose to move the underlying comparison functions to `docker_common.py` to be able to use them in other modules as well.

As a first example, I've updated the docker_config and docker_secret modules to use it for comparing labels. As you can see, this simplified their idempotency checks a lot :)

Other modules which can benefit are docker_network and docker_volume. (Changes to docker_network will create conflicts with #47492, so we'll have to coordinate with that PR.) For both modules, note that they return differences in a different format than docker_container (they list effectively different parameters with subelements; while docker_container provides a list of different parameters together with the actual values, so the user can figure out herself which part differs if it is a dict/list/...). @DBendit @agronholm any opinions on this?

##### ISSUE TYPE
- Bugfix Pull Request

(It is not really a bugfix PR, but the other categories fit even less...)

##### COMPONENT NAME
lib/ansible/module_utils/docker_commons.py
docker_config
docker_configuration
docker_secret

##### ANSIBLE VERSION
```
2.8.0
```

felixfontein committed:
More generic comparison code from docker_container to docker_common.

felixfontein committed:
More flexibility if a is None and method is allow_to_present.

Note that this odes not affect docker_container, as there a is never None.

felixfontein committed:
Update docker_secret and docker_config: simplify labels comparison.

felixfontein commented:
Moved 16f058811b5464cce0e99ad938f022d158b7c723 to #47712.

DBendit commented:
I like this quite a bit. For one, consolidating comparison logic means that the output for all (eventually, anyway) docker modules can be consistent, which I think is a good goal. Two, I prefer the output style of docker_container. Presenting the parameter alongside the present value makes it much clearer as to why something's changing without requiring any further checks by the user.

I'll have to give the code another pass later, but I approve of this idea conceptually.

felixfontein commented:
Rebased to account for #47712 being merged.

DBendit commented on lib/ansible/module_utils/docker_common.py:
```diff
@@ -541,3 +541,103 @@ def pull_image(self, name, tag="latest"):
+    return True
+
+
+def compare_generic(a, b, method, type):
```
I recognize that this will be covered by the existing integration tests, but would it be worthwhile to add unit tests around this functionality? There's enough logic in here that I think it merits additional validation.

felixfontein replied:
A very good point! I'll add something.

felixfontein committed:
Added unit tests.

felixfontein committed:
Use proper subsequence test for allow_more_present for lists.

Note that this does not affect existing code in docker_container, since lists
don't use allow_more_present. Using allow_more_present will only be possible
in Ansible 2.8.

felixfontein commented:
I've added some unit tests. I also changed the behavior for `list`s with `allow_more_present`. Before, comparison as sets was used in that case; now it does a proper subsequence test. This doesn't affect `docker_container` since the combination `list`+`allow_more_present` will only be possible with Ansible 2.8 (using the new `comparisons` option).

felixfontein committed:
pep8

felixfontein commented:
Restarting tests.

DBendit commented:
LGTM

Perfectly fine with moving the rest of the modules to this comparison method in a future PR. Might be worth writing up a quick Issue for it so it doesn't get dropped?

kassiansun reviewed (approved):
shipit

gundalow commented:
@felixfontein thank you for the PR and tests
@DBendit thank you for review and suggestion of unit tests. 

Nice to see PRs that refactor code

felixfontein commented:
@kassiansun @DBendit thanks for the reviews!
@gundalow thanks for merging!

I'll create backport PRs for this.
